src/modules/clustering/hierarchical_clustering_method.cpp

```diff
@@ -0,0 +1,169 @@
+}
+
+
+// Partition based on mate2
+std::vector<std::vector<Junction>> split_partition_based_on_mate2(std::vector<Junction> const & partition)
+{
+    std::vector<Junction> current_partition{};
+    std::vector<std::vector<Junction>> splitted_partition{};
+
+    for (Junction junction : partition)
+    {
+        if (current_partition.empty())
+        {
+            current_partition.push_back(junction);
+        }
+        else
+        {
+            if (junction.get_mate2().seq_name != current_partition.back().get_mate2().seq_name ||
+                junction.get_mate2().orientation != current_partition.back().get_mate2().orientation ||
+                abs(junction.get_mate2().position - current_partition.back().get_mate2().position) > 50)
+            {
+                std::sort(current_partition.begin(), current_partition.end());
+                splitted_partition.push_back(current_partition);
+                current_partition.clear();
+            }
+            current_partition.push_back(junction);
+        }
+    }
+    if (!current_partition.empty())
+    {
+        std::sort(current_partition.begin(), current_partition.end());
+        splitted_partition.push_back(current_partition);
+    }    
+    return splitted_partition;
+}
+
+int junction_distance(Junction const & lhs, Junction const & rhs)
+{
+    if ((lhs.get_mate1().seq_name == rhs.get_mate1().seq_name) &&
+        (lhs.get_mate1().orientation == rhs.get_mate1().orientation) &&
+        (lhs.get_mate2().seq_name == rhs.get_mate2().seq_name) &&
+        (lhs.get_mate2().orientation == rhs.get_mate2().orientation))
+    {
+        return (std::abs(lhs.get_mate1().position - rhs.get_mate1().position) +
+                std::abs(lhs.get_mate2().position - rhs.get_mate2().position) +
+                std::abs((int)(lhs.get_inserted_sequence().size() - rhs.get_inserted_sequence().size())));
+    }
+    else
+    {
+        return std::numeric_limits<int>::infinity();
+    }        
+}
+
+std::vector<Cluster> hierarchical_clustering_method(std::vector<Junction> const & junctions,
+                                                    double clustering_cutoff)
+{
+    auto partitions = partition_junctions(junctions);
+    std::vector<Cluster> clusters{};
+    for (std::vector<Junction> partition : partitions)
+    {
+        size_t partition_size = partition.size();
+        if (partition_size < 2)
+        {
+            clusters.emplace_back(partition);
+            continue;
+        }
+        // Compute condensed distance matrix (upper triangle of the full distance matrix)
+        double* distmat = new double[(partition_size * (partition_size - 1)) / 2];
+        int k, i, j;
+        for (i = k = 0; i < partition_size; ++i) {
+            for (j = i + 1; j< partition_size; ++j) {
+                // Compute distance between junctions i and j  
+                distmat[k] = junction_distance(partition[i], partition[j]);
+                ++k;
+            }
+        }
```

Comment:
- Better prevent some leakage by defining variables inside the for loop.
- More consistency by using `size_t`.
- You can have multiple operation by seperating them by a comma: `++j, ++k`
```suggestion
        for (size_t i{0}, k{0}; i < partition_size; ++i) {
            for (size_t j{i + 1}; j< partition_size; ++j, ++k) {
                // Compute distance between junctions i and j  
                distmat[k] = junction_distance(partition[i], partition[j]);
            }
        }
```